keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,534 @@
+updates can potentially stress the kube-apiserver. For this reason, the Job
+controller repeats steps 2 to 4, capping the number of Pods each time, until
+the controller processes all the Pods. The number of Pods is caped by:
+- time: in each iteration, the job controller removes all the Pods' finalizers
+  it can in 1 minute. This allows to throttle the number of status updates.
+- count: Preventing big writes to etcd.
+
+### User deleted Pods
+   
+In the case where a user or another controller removes a Pod, which sets a
+deletion timestamp, the Job controller removes the finalizer from the Pod,
+<<[UNRESOLVED: should we count terminating pods?]>>
+independently of it having finished or not.
+This allows for deleted Pods to not count towards the failed counter, so it
+doesn't change the current behavior.
+Note that step 2 doesn't pay attention to deletion timestamps. So the Job
+controller might account the Pod in the lists and counters if the Pods
+enter Failed or Succeeded phase before or while being deleted.
+<<[/UNRESOLVED]>>
+   
+### Pod adoption
+
+If a Job with `.status.uncountedTerminatedPodUIDs.enabled = true` can adopt a Pod
+(according to the existing adoption criteria), this Pod might not have a
+finalizer.
+
+The job controller adds the finalizer in the same patch request that modifies
+the owner reference.
+
+### Test Plan
+
+- Unit tests:
+  - Job sync with feature gate enabled.
+  - Removal of finalizers when feature gate is disabled.
+  - Tracking of terminating Pods.
+- Integration tests:
+  - Job tracking with feature enabled.
+  - Tracking of terminating Pods.
+  - Transition from feature enabled to disabled and enabled again.
+- E2E test:
+  - Job tracking with feature enabled.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implementation:
+  - Job tracking without lingering Pods
+  - Removal of finalizer when feature gate is disabled.
+- Tests: unit, integration, E2E
+
+#### Alpha -> Beta Graduation
+
+- Implementation:
+  - Index tracking for [Indexed Jobs](http://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job)
+- Metrics:
+  - latency
+  - errors
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- E2E test graduates to conformance.
+- Job tracking scales to 10^5 completions.
+
+### Upgrade / Downgrade Strategy
+
+When the feature `JobTrackingWithoutLingeringPods` is enabled for the first
+time, the cluster can have Jobs whose Pods don't have the
+`batch.kubernetes.io/job-completion` finalizer. It would be hard to add the
+finalizer to all Pods while preventing race conditions.
+
+We use `.status.uncountedTerminatedPodUIDs.enabled` to indicate whether the Job
+was created after the feature was enabled. If this boolean is false, the Job
+controller tracks Pods using the legacy algorithm.
+
+The kube-apiserver sets `.status.uncountedTerminatedPodUIDs.enabled = true` when
+the feature gate `JobTrackingWithoutLingeringPods` is enabled, at Job creation.
```

Comment:
I'm not sure if this is actually needed, the controller should be capable of reacting to the current state of the job and update its status accordingly.

Reply:
We need this for any new API field to be able to handle partial upgrades in a HA control plane.